### PR TITLE
Use `with` blocks for closing and opening connections

### LINKED ISSUE #337: Socket not closable before program end.
When doing something like 
```
server = JavaServer.lookup("example.com:25565")
status = server.status()
```
There does not appear to be a way to close the socket connection (normally done on program end). Because of this, large amounts of calls to get server data from a queue will run into the error "Too many open files", the files being the sockets. The only way I can currently see to close the socket is to end the program. If there is another way to do this, please correct me.

### PR DESCRIPTION
**Pull request by PerchunPak**
Fixes #337.

### CODE REVIEW

File: mcstatus/protocol/connection.py [hunk 1]
```diff
@@ -512,12 +512,11 @@ def close(self) -> None:
...
-    def __del__(self) -> None:
-        """Close self.socket."""
-        try:
-            self.close()
-        except OSError:  # Probably, the socket was already closed by the OS
-            pass
+    def __enter__(self) -> Self:
+        return self
+
+    def __exit__(self, _, __, ___) -> None:  # noqa: ANN001
```
Comment (ItsDrike) [inline mcstatus/protocol/connection.py:518]:
```suggestion
    def __exit__(self, *_args) -> None:
```

File: mcstatus/protocol/connection.py [hunk 2]
```diff
@@ -623,30 +622,30 @@ def write(self, data: Connection | str | bytes | bytearray) -> None:
...
-                pass
+            self.writer.close()
+
+    async def __aenter__(self) -> Self:
+        await self.connect()
+        return self
 
-    def __del__(self) -> None:
-        """Close self."""
+    async def __aexit__(self, _, __, ___) -> None:  # noqa: ANN001
```
Comment (ItsDrike) [inline mcstatus/protocol/connection.py:631]:
```suggestion
    async def __aexit__(self, *_args) -> None:
```

File: mcstatus/protocol/connection.py [hunk 3]
```diff
@@ -671,9 +670,9 @@ def close(self) -> None:
...
-        """Close self.stream"""
-        try:
-            self.close()
-        except (asyncio.exceptions.CancelledError, asyncio.exceptions.TimeoutError):
-            return
+    async def __aenter__(self) -> Self:
+        await self.connect()
+        return self
+
+    async def __aexit__(self, _, __, ___) -> None:  # noqa: ANN001
```
Comment (ItsDrike) [inline mcstatus/protocol/connection.py:677]:
```suggestion
    async def __aexit__(self, *_args) -> None:
```

File: tests/test_server.py [hunk 4]
```diff
@@ -102,21 +102,19 @@ def test_ping(self):
...
 
     def test_ping_retry(self):
-        with patch("mcstatus.server.TCPSocketConnection") as connection:
-            connection.return_value = None
-            with patch("mcstatus.server.ServerPinger") as pinger:
-                pinger.side_effect = [Exception, Exception, Exception]
-                with pytest.raises(Exception):
-                    self.server.ping()
-                assert pinger.call_count == 3
+        with patch("mcstatus.server.TCPSocketConnection"), patch("mcstatus.server.ServerPinger") as pinger:
```
Comment (ItsDrike) [inline tests/test_server.py:113]:
```suggestion
        # Use a blank mock for the connection, we don't want to actually create any connections
        with patch("mcstatus.server.TCPSocketConnection"), patch("mcstatus.server.ServerPinger") as pinger:
```

File: tests/test_server.py [hunk 5]
```diff
@@ -141,13 +139,11 @@ def test_status(self):
...
 
     def test_status_retry(self):
-        with patch("mcstatus.server.TCPSocketConnection") as connection:
-            connection.return_value = None
-            with patch("mcstatus.server.ServerPinger") as pinger:
-                pinger.side_effect = [Exception, Exception, Exception]
-                with pytest.raises(Exception):
-                    self.server.status()
-                assert pinger.call_count == 3
+        with patch("mcstatus.server.TCPSocketConnection"), patch("mcstatus.server.ServerPinger") as pinger:
```
Comment (ItsDrike) [inline tests/test_server.py:142]:
```suggestion
        # Use a blank mock for the connection, we don't want to actually create any connections
        with patch("mcstatus.server.TCPSocketConnection"), patch("mcstatus.server.ServerPinger") as pinger:
```

File: tests/test_server.py [hunk 6]
```diff
@@ -187,14 +183,12 @@ def test_query(self):
...
     def test_query_retry(self):
-        with patch("mcstatus.server.UDPSocketConnection") as connection:
-            connection.return_value = None
-            with patch("mcstatus.server.ServerQuerier") as querier:
-                querier.side_effect = [Exception, Exception, Exception]
-                with pytest.raises(Exception), patch.object(self.server.address, "resolve_ip") as resolve_ip:
-                    resolve_ip.return_value = "127.0.0.1"
-                    self.server.query()
-                assert querier.call_count == 3
+        with patch("mcstatus.server.UDPSocketConnection"), patch("mcstatus.server.ServerQuerier") as querier:
```
Comment (ItsDrike) [inline tests/test_server.py:186]:
```suggestion
        # Use a blank mock for the connection, we don't want to actually create any connections
        with patch("mcstatus.server.UDPSocketConnection"), patch("mcstatus.server.ServerQuerier") as querier:
```

**Review by ItsDrike — CHANGES_REQUESTED** [review]
Good work! Just some minor things and it should be good to go

File: mcstatus/protocol/connection.py (hunk 1 above)
Comment (PerchunPak) [reply mcstatus/protocol/connection.py]: Fixed [here](https://github.com/py-mine/mcstatus/compare/a9db60192fd04768826160a255837ca4b4ebb628..76137712f61712569cee5887cdacbaba921e96f0).

File: mcstatus/protocol/connection.py (hunk 2 above)
Comment (PerchunPak) [reply mcstatus/protocol/connection.py]: Fixed [here](https://github.com/py-mine/mcstatus/compare/a9db60192fd04768826160a255837ca4b4ebb628..76137712f61712569cee5887cdacbaba921e96f0).

File: mcstatus/protocol/connection.py (hunk 3 above)
Comment (PerchunPak) [reply mcstatus/protocol/connection.py]: Fixed [here](https://github.com/py-mine/mcstatus/compare/a9db60192fd04768826160a255837ca4b4ebb628..76137712f61712569cee5887cdacbaba921e96f0).

File: tests/test_server.py (hunk 5 above)
Comment (PerchunPak) [reply tests/test_server.py:144]: Fixed [here](https://github.com/py-mine/mcstatus/compare/a9db60192fd04768826160a255837ca4b4ebb628..76137712f61712569cee5887cdacbaba921e96f0).

File: tests/test_server.py (hunk 6 above)
Comment (PerchunPak) [reply tests/test_server.py:189]: Fixed [here](https://github.com/py-mine/mcstatus/compare/a9db60192fd04768826160a255837ca4b4ebb628..76137712f61712569cee5887cdacbaba921e96f0).

**Pull request merged**